packages/core/content-manager/server/src/controllers/relations.ts

```diff
@@ -45,117 +44,199 @@ const sanitizeMainField = (model: any, mainField: any, userAbility: any) => {
+      ? await getService('components').findConfiguration(sourceSchema)
+      : await getService('content-types').findConfiguration(sourceSchema);
 
-    const modelConfig = isComponent
-      ? await getService('components').findConfiguration(modelSchema)
-      : await getService('content-types').findConfiguration(modelSchema);
+    const targetSchema = strapi.getModel(attribute.target);
 
     const mainField = flow(
       prop(`metadatas.${targetField}.edit.mainField`),
       (mainField) => mainField || 'id',
-      (mainField) => sanitizeMainField(targetedModel, mainField, userAbility)
+      (mainField) => sanitizeMainField(targetSchema, mainField, userAbility)
     )(modelConfig);
 
-    const fieldsToSelect = uniq(['id', mainField, PUBLISHED_AT_ATTRIBUTE]);
+    const fieldsToSelect = uniq([mainField, PUBLISHED_AT_ATTRIBUTE, 'documentId']);
+    if (locale) {
+      fieldsToSelect.push('locale');
+    }
+
+    return {
+      attribute,
+      fieldsToSelect,
+      mainField,
+      sourceSchema,
+      targetSchema,
+      targetField,
+      currentEntityId: currentEntity?.id,
+    };
+  },
+
+  async findAvailable(ctx: any) {
+    await validateFindAvailable(ctx.request.query);
+
+    const { id } = ctx.request.query;
+    const locale = ctx.request?.query?.locale || null;
+    const status = ctx.request?.query?.status || 'draft';
+
+    const validation = await this.extractAndValidateRequestInfo(ctx, status, id, locale);
+    if (!validation) {
+      // If validation of the request has failed the error has already been sent
+      // to the ctx
+      return;
+    }
+
+    const {
+      targetField,
+      fieldsToSelect,
+      mainField,
+      sourceSchema: { uid: sourceUid },
+      targetSchema: { uid: targetUid },
+      currentEntityId,
+    } = validation;
+
+    const { idsToOmit, idsToInclude, _q, ...query } = ctx.request.query;
 
     const permissionChecker = getService('permission-checker').create({
-      userAbility,
-      model: targetedModel.uid,
+      userAbility: ctx.state.userAbility,
+      model: targetUid,
     });
     const permissionQuery = await permissionChecker.sanitizedQuery.read(query);
 
     const queryParams = {
       sort: mainField,
-      fields: fieldsToSelect, // cannot select other fields as the user may not have the permissions
+      // cannot select other fields as the user may not have the permissions
+      fields: fieldsToSelect,
       ...permissionQuery,
     };
 
-    if (!isEmpty(idsToOmit)) {
-      addFiltersClause(queryParams, { id: { $notIn: idsToOmit } });
+    // Only ever findAvailable (e.g. populate relation select) on a draft entry
+    addFiltersClause(queryParams, {
+      [PUBLISHED_AT_ATTRIBUTE]: null,
+    });
+
+    // We are looking for available content type relations and should be
+    // filtering by valid documentIds only
+    const stringIdsToOmit = idsToOmit?.filter((id: any) => !Number(id));
+    if (stringIdsToOmit?.length > 0) {
```

Comment:
won't all idsToOmit be document ids?

Reply:
I think you're right as a target will never be a component, I will check

Reply:
was this the case?